### DISCUSSION TITLE
`task.kwargs` from events is truncated for long kwargs

### DISCUSSION THREAD

**Discussion by galbiati** [celery/celery#7777] [Q&A]
Hi!

I'm trying to write some [realtime event handlers](https://docs.celeryq.dev/en/stable/userguide/monitoring.html#real-time-processing) that require information from task kwargs. 

However, the kwargs get truncated when they are too long. This means they can't be parsed normally as JSON, and sometimes the information I need gets cut off.

Short of just making kwargs shorter (very not ideal for me), is there a way to configure the maximum character length for the kwargs, or something similar? And would this setting be applied at the receiver, or is it something that would require restarting workers?

**Comment by think-divergent** [comment]
Here's small open PR that might be relevant: https://github.com/celery/celery/pull/6595 
Monkey patching celery.utils.saferepr.saferepr on the receiver at run time could potentially be a short term workaround.

**Reply by galbiati** [reply]
Ah nice; thanks for the reference! (Was having a bad time figuring out what to even search for).

Will let you know if I can use a patch successfully. PR might not work for me as-is; looks like it targets ampq but not redis? Might look into proposing an expansion to the pr if that's the case...

But even having a starting point is miles ahead of where I was a few hours ago - thanks again 🤙

**Reply by galbiati** [reply]
Hm, it looks like `saferepr` [is being used](https://github.com/celery/celery/blob/ad994719bafe6747af6cf8251efb0925284a9260/celery/worker/strategy.py#L84) to send [kwargs](https://github.com/celery/celery/blob/ad994719bafe6747af6cf8251efb0925284a9260/celery/worker/strategy.py#L168) for events. So I think they're truncated when they're sent unfortunately, and patching `saferepr` would have to be done in the workers.

Still unsure how this actually happens. `saferepr` doesn't have a default `maxlen`, and on those requests/strategy objects `saferepr` doesn't pass anything. 

For AMPQ, this looks like it gets used [here](https://github.com/celery/celery/blob/1c4ff33bd22cf94e297bd6449a06b5a30c2c1fbc/celery/app/amqp.py#L319). The default value on the `AMPQ` class matches what I see (truncated to 1023 chars). But I haven't figured out how that's getting applied to my case with redis...

**Reply by think-divergent** [reply]
AMPQ might not be the best name here. It seems to run on an abstracted [producer pool](https://github.com/celery/celery/blob/1c4ff33bd22cf94e297bd6449a06b5a30c2c1fbc/celery/app/amqp.py#L590) with [kombu](https://github.com/celery/celery/blob/1c4ff33bd22cf94e297bd6449a06b5a30c2c1fbc/celery/app/amqp.py#L8) handling the queue specific implementations.

**Reply by galbiati** [reply]
Got it; yeah that makes more sense. Thanks for the assist here - doesn't quite solve my problem, but at least I understand what's going on and why I'll have to change my design ;)
